src/hardware/uni-t-ut8802e/protocol.c

```diff
@@ -0,0 +1,430 @@
+            info->meas_data.main_unit = SR_UNIT_UNITLESS;
+            break;
+
+        case MODE_THYRISTOR_SCR:
+            info->meas_head.mode = SR_MQ_VOLTAGE;
+            info->meas_data.main_unit = SR_UNIT_VOLT;
+            break;
+
+        case MODE_FREQ_Hz:
+            info->meas_head.mode = SR_MQ_FREQUENCY;
+            info->meas_data.main_unit = SR_UNIT_HERTZ;
+            info->meas_data.main_prec = range_hz[0];
+            break;
+        case MODE_FREQ_KHz:
+            info->meas_head.mode = SR_MQ_FREQUENCY;
+            info->meas_data.main_unit = SR_UNIT_HERTZ;
+            info->meas_data.main_prec = range_hz[1];
+            break;
+        case MODE_FREQ_MHz:
+            info->meas_head.mode = SR_MQ_FREQUENCY;
+            info->meas_data.main_unit = SR_UNIT_HERTZ;
+            info->meas_data.main_prec = range_hz[2];
+            break;
+
+        case MODE_DUTY:
+            info->meas_head.mode = SR_MQ_DUTY_CYCLE;
+            info->meas_data.main_unit = SR_UNIT_PERCENTAGE;
+            break;
+
+        default:
+            sr_spew("Unkwon functionality");
+            return SR_ERR_DATA;
+    }
+
+    ret = ut8802e_feedbuff_initialization(&feedbuff);
+    if (ret != SR_OK)
+        return SR_ERR_DATA;
+
+    g_slist_free(feedbuff.analog.meaning->channels);
+    feedbuff.analog.meaning->channels = g_slist_append(
+        NULL,
+        g_slist_nth_data(sdi->channels, ut8802e_CH_MAIN));
+
+    feedbuff.analog.meaning->mqflags = info->meas_head.mqflag;
+    feedbuff.analog.meaning->mq = info->meas_head.mode;
+    feedbuff.analog.meaning->unit = info->meas_data.main_unit;
+
+    info->meas_data.main_value = new_value * pow(10, info->meas_data.main_prec);
+
+    feedbuff.analog.encoding->digits = -info->meas_data.main_prec + 4;
+    feedbuff.analog.spec->spec_digits = -info->meas_data.main_prec + 4;
+
+    feedbuff.main_value = info->meas_data.main_value;
+	if (sdi->status != SR_ST_ACTIVE)
+		return SR_OK;
+
+    ret = sr_session_send(sdi, &feedbuff.packet);
+    if (ret != SR_OK)
+        return SR_ERR_DATA;
+
+    sr_sw_limits_update_samples_read(&devc->limits, 1);
+    if (sr_sw_limits_check(&devc->limits))
+        sr_dev_acquisition_stop(sdi);
+
+    if (feedbuff.analog.meaning)
+        g_slist_free(feedbuff.analog.meaning->channels);
+
+    return SR_OK;
+}
+
+static int process_buffer(struct sr_dev_inst *sdi) {
+    struct dev_context *devc;
+    uint8_t *pkt;
+    uint8_t v8;
+    size_t pkt_len, remain, idx;
+    int ret;
+    GString *spew;
+
+    devc = sdi->priv;
+    pkt = &devc->packet[0];
+
+    do {
+        if (devc->packet_len < 1 * sizeof(uint8_t))
+            return SR_OK;
+
+        v8 = R8(&pkt[0]);
+        if (v8 != FRAME_MAGIC)
+            break;
+
+        pkt_len = 8;
+        sr_spew("Got a expected packet lenght %zu, have %zu",
+                pkt_len, devc->packet_len);
+
+        if (pkt_len > devc->packet_len)
+            return SR_OK;
```

Comment:
Surely there's no need to check packet validity both here and in `process_packet()`?